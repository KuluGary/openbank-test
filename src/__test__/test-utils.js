import React from "react"
import { render as rtlRender } from "@testing-library/react"
import { Provider } from "react-redux"

import reduxStore from "store";

/**
 * Wraps the component in a mock of the redux store provider
 * @param {Component} ui - the UI component passed on render
 * @param {Object} renderOptions - the render options passed 
 */
const render = (ui, { initialState, store = reduxStore, ...renderOptions } = {}) => {
    const Wrapper = ({ children }) => {
        return <Provider store={store}>{children}</Provider>
    }

    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from "@testing-library/react"; // export full testing-library components
export { render }; // export provider wrapped render
