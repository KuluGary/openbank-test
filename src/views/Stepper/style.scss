@import "../../styles/base";

div.stepper {
    background-color: #CCD5DA;
    min-height: 6rem;

    &-container {
        background: transparent;
        max-width: 40%;
        margin: 0 auto;
    }

    .MuiStepIcon-root {
        color: #788F9C;
    }
    
    .MuiStepIcon-completed {
        color: $primary-color;
    }
    
    .MuiStepIcon-active {
        color: $secondary-color;
        width: 1.5em;
        height: 1.5em;
        transition: width .2s, height .2s;
    }
    

    .MuiStep-horizontal {
        padding: 0;

        .MuiStepLabel-iconContainer {
            padding: 0
        }
    }

    .MuiStepConnector-line {
        border-top-width: 3px;
    }

    .Mui-disabled {        
        .MuiStepConnector-line {
            border-color: #788F9C;
        }
    }

    .MuiStepConnector-active {        
        .MuiStepConnector-line {
            border-top-width: 3px;
            border-color: $primary-color;
        }
    }

    .MuiStepConnector-completed {        
        .MuiStepConnector-line {
            border-top-width: 3px;
            border-color: $primary-color;
        }
    }
}