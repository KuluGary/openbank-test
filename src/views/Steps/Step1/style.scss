.step1 {
    padding: 1.5rem 3rem;
    
    &-wrapper {
        display: flex;
        justify-content: space-around;
        flex-wrap: wrap;

        @media (max-width: 768px) {
            display: block;
            margin: 0 auto;
        }

        &-image-container {
            max-width: 35%;
            text-align: justify;

            @media (max-width: 768px) {
                max-width: 75%;
                margin: 0 auto;
            }

            img {
                display: block;
                margin: 0 auto;
            }
        }
    }
}