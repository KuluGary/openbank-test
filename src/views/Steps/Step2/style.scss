.step2 {
    padding: 1.5rem 3rem;

    &-wrapper {
        &:last-child {
            margin: 3rem 0;
        }
        &-flex {
            max-width: 90%;
            display: flex;
            @media (max-width: 1020px) {
                display: block;
                margin: 0 auto;
            }
        }

        &-field {
            @media (max-width: 1020px) {
                width: 100%;
            }
            width: 35%;

            &:first-of-type {
                margin-right: 2rem;
            }
        }
    }
}