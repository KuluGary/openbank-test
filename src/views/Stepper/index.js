import React from 'react'
import PropTypes from 'prop-types'
import { Stepper as MUIStepper, Step, StepLabel } from '@material-ui/core'
import './style.scss'

/**
 * Component that creates and controls the steps
 * Uses Material UI Stepper component
 * @param {Number} activeStep - current index of active step
 * @param {Array} step - Array of step components
 */
const Stepper = ({ activeStep, steps }) => (
    <div className="stepper" data-testid="stepper">
        <MUIStepper
            activeStep={activeStep}
            classes={{ root: 'stepper-container' }}>
            {steps.map((_, index) => (
                <Step
                    key={index}
                    data-testid="step">
                    <StepLabel />
                </Step>
            ))}
        </MUIStepper>
    </div>
)

Stepper.propTypes = {
  activeStep: PropTypes.number,
  steps: PropTypes.array
}

export default Stepper
