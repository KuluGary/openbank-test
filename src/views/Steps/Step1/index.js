import React from "react";
import ProtectPassword from "assets/img/group.svg";
import MasterKey from "assets/img/group-3.svg";
import { useTranslation } from "react-i18next";
import "./style.scss";
import { Checkbox, FormControlLabel } from "@material-ui/core";

/**
 * Function component for rendering the first step
 * of the wizard
 */
const Step1 = ({ validator, setValidator }) => {
  const { t } = useTranslation();
  
  // Set default state of current validator to false
  React.useEffect(() => setValidator(false), [setValidator]);

  const handleCheckChange = () => {
    setValidator(!validator);
  };

  return (
    <div className="step1" data-testid="step1">
      <h1>{t("wizard.header")}</h1>
      <div className="step1-wrapper">
        <div className="step1-wrapper-image-container">
          <img alt="Protect your password" src={ProtectPassword} />
          <p>{t("wizard.step1.protect_password")}</p>
        </div>
        <div className="step1-wrapper-image-container">
          <img alt="Protect your password" src={MasterKey} />
          <p>{t("wizard.step1.master_key")}</p>
        </div>
      </div>
      <h4>{t("wizard.step1.how_it_works_header")}</h4>
      <p>{t("wizard.step1.how_it_works")}</p>

      <h4>{t("wizard.step1.which_data_header")}</h4>
      <p>{t("wizard.step1.which_data")}</p>
      <div className="step1-checkbox">
        <FormControlLabel
          control={
            <Checkbox
              checked={validator}
              onChange={handleCheckChange}
              inputProps={{
                "data-testid": "privacy-checkbox",
              }}
            />
          }
          label={t("wizard.step1.checkbox")}
        />
      </div>
    </div>
  );
};

export default Step1;
