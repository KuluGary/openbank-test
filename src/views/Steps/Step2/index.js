import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { TextField, IconButton, InputAdornment } from '@material-ui/core'
import { Visibility, VisibilityOff } from '@material-ui/icons'
import { checkPassword } from 'services/string-util';
import { useTranslation } from 'react-i18next'
import {
  setPassword, setPasswordHint, setPasswordError
} from 'store/password'
import './style.scss'

/**
 * Function component for rendering the first step
 * of the wizard
 */
export default function Step2 ({ setValidator }) {
  const dispatch = useDispatch()
  const { t } = useTranslation()
  const {
    password,
    passwordHint,
    passwordError
  } = useSelector(state => state.password)
  const [showPassword, setShowPassword] = React.useState(false)
  const [passwordCheck, setPasswordCheck] = React.useState('')
  const maxPasswordHintLength = 255

  // Set default state of current validator to false
  React.useEffect(() => setValidator(false), [setValidator])

  // Error handling
  React.useEffect(() => {
    if (password.length > 0 || passwordCheck.length > 0) {
      let errorMessage = null
      let error = false;

      // Manage validation for passwords
      if (!checkPassword(password)) {
        errorMessage = t('wizard.step2.pass_error_regex')
        error = true
      } else if (!passwordCheck) {
        errorMessage = t('wizard.step2.pass_error_match')
        error = true
      } else if (password !== passwordCheck) {
        errorMessage = t('wizard.step2.pass_error_match')
        error = true
      }
      
      // Manage validation for hints
      if (passwordHint.length > maxPasswordHintLength) {
        error = true
      }

      dispatch(setPasswordError(errorMessage))
      setValidator(!error)
    }
  }, [password, passwordCheck, dispatch, t, setValidator, passwordHint])

  const handlePasswordChange = (e) => dispatch(setPassword(e.target.value))

  const handleHintChange = (e) => dispatch(setPasswordHint(e.target.value))

  return (
    <div className="step2" data-testid="step2">
        <h1>{t('wizard.header')}</h1>
        <div className="step2-wrapper">
            <p>{t('wizard.step1.how_it_works')}</p>
            <div className="step2-wrapper-flex">
                <div className="step2-wrapper-field">
                    <TextField
                        fullWidth
                        error={!!passwordError}
                        label={t('wizard.step2.create_pass_label')}
                        variant="outlined"
                        color="secondary"
                        type={showPassword ? 'text' : 'password'}
                        onChange={handlePasswordChange}
                        helperText={passwordError || ' '}
                        InputProps={{
                          inputProps: {
                            "data-testid": "password-input"
                          },
                          endAdornment: <InputAdornment position="end">
                                <IconButton
                                    onClick={() => setShowPassword(!showPassword)}
                                    aria-label="toggle password visibility">
                                    {showPassword ? <VisibilityOff /> : <Visibility />}
                                </IconButton>
                            </InputAdornment>
                        }} />
                </div>
                <div className="step2-wrapper-field">
                    <TextField
                        fullWidth
                        label={t('wizard.step2.repeat_pass')}
                        variant="outlined"
                        color="secondary"
                        type={showPassword ? 'text' : 'password'}
                        onChange={(e) => setPasswordCheck(e.target.value)}
                        InputProps={{
                          inputProps: {
                            "data-testid": "password-repeat-input"
                          },
                          endAdornment: <InputAdornment position="end">
                                <IconButton
                                    onClick={() => setShowPassword(!showPassword)}
                                    aria-label="toggle password visibility">
                                    {showPassword ? <VisibilityOff /> : <Visibility />}
                                </IconButton>
                            </InputAdornment>
                        }} />
                </div>
            </div>
        </div>
        <div className="step2-wrapper">
            <p>{t('wizard.step2.hint_explanation')}</p>
            <TextField
                fullWidth
                variant="outlined"
                color="secondary"
                value={passwordHint}
                error={passwordHint.length >= maxPasswordHintLength}
                helperText={`${passwordHint.length}/${maxPasswordHintLength}`}
                label={t('wizard.step2.hint_label')}
                onChange={handleHintChange}
            />
        </div>
    </div>
  )
}
