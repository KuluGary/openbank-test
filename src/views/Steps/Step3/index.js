import React from 'react'
import { useSelector } from 'react-redux'
import { useTranslation } from 'react-i18next'
import { CircularProgress } from '@material-ui/core'
import { submitForm } from 'services/api'
import success from 'assets/img/checked.svg'
import warning from 'assets/img/warning.svg'
import './style.scss'

/**
 * Function component for rendering the first step
 * of the wizard
 */
const Step3 = () => {
  const { t } = useTranslation()
  const [loading, setLoading] = React.useState(true)
  const [status, setStatus] = React.useState(200)
  const { password, passwordHint } = useSelector(state => state.password)

  React.useEffect(() => {
    submitForm(password, passwordHint)
      .then(({ status }) => {
        setStatus(status)
        setLoading(false)
      })
      .catch(({ status }) => {
        setStatus(status)
        setLoading(false)
      })
  }, [password, passwordHint])

  if (loading) {
    return (
            <div className="step3">
                <CircularProgress
                    color="secondary" />
            </div>
    )
  }

  if (status > 300) {
    return (
            <div className="step3">
                <div className="step3-wrapper">
                    <div className="step3-wrapper-image-container">
                        <img alt="Protect your password" src={warning} />
                        <div className="step3-wrapper-image-container-text">
                            <h2>{t('wizard.step3.warning_header')}</h2>
                            <p>{t('wizard.step3.warning')} </p>
                        </div>
                    </div>
                </div>
            </div>
    )
  }

  return (
        <div className="step3">
            <div className="step3-wrapper">
                <div className="step3-wrapper-image-container">
                    <img alt="Protect your password" src={success} />
                    <div className="step3-wrapper-image-container-text">
                        <h2>{t('wizard.step3.success_header')}</h2>
                        <p>{t('wizard.step3.success')}</p>
                    </div>
                </div>
            </div>
        </div>
  )
}

export default Step3
