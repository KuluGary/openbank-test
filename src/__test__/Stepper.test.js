import React from "react";
import Stepper from "views/Stepper";
import Step1 from "views/Steps/Step1";
import Step2 from "views/Steps/Step2";
import Step3 from "views/Steps/Step3";

import { render, cleanup } from "./test-utils";
import "@testing-library/jest-dom/extend-expect";

describe("Stepper", () => {
    afterEach(cleanup);
    const steps = [Step1, Step2, Step3];
    const activeStep = 0;

    it("matches snapshot", () => {
        const div = document.createElement("div");
        const component = render(<Stepper steps={steps} activeStep={activeStep} />, div);
        expect(component.container).toMatchSnapshot();
    })

    it("renders component", () => {
        const { getByTestId } = render(<Stepper steps={steps} activeStep={activeStep} />);

        expect(getByTestId("stepper")).toBeInTheDocument();
    })

    it("renders the right amount of steps", () => {
        const { getAllByTestId } = render(<Stepper steps={steps} activeStep={activeStep} />);

        expect(getAllByTestId("step").length).toEqual(steps.length);            
    })
})