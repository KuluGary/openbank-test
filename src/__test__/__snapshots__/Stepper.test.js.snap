// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper matches snapshot 1`] = `
<div>
  <div
    class="stepper"
    data-testid="stepper"
  >
    <div
      class="MuiPaper-root MuiStepper-root stepper-container MuiStepper-horizontal MuiPaper-elevation0"
    >
      <div
        class="MuiStep-root MuiStep-horizontal"
        data-testid="step"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-horizontal"
        >
          <span
            class="MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiStepIcon-root MuiStepIcon-active"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                class="MuiStepIcon-text"
                text-anchor="middle"
                x="12"
                y="16"
              >
                1
              </text>
            </svg>
          </span>
          <span
            class="MuiStepLabel-labelContainer"
          />
        </span>
      </div>
      <div
        class="MuiStepConnector-root MuiStepConnector-horizontal Mui-disabled"
      >
        <span
          class="MuiStepConnector-line MuiStepConnector-lineHorizontal"
        />
      </div>
      <div
        class="MuiStep-root MuiStep-horizontal"
        data-testid="step"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled"
        >
          <span
            class="MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiStepIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                class="MuiStepIcon-text"
                text-anchor="middle"
                x="12"
                y="16"
              >
                2
              </text>
            </svg>
          </span>
          <span
            class="MuiStepLabel-labelContainer"
          />
        </span>
      </div>
      <div
        class="MuiStepConnector-root MuiStepConnector-horizontal Mui-disabled"
      >
        <span
          class="MuiStepConnector-line MuiStepConnector-lineHorizontal"
        />
      </div>
      <div
        class="MuiStep-root MuiStep-horizontal"
        data-testid="step"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled"
        >
          <span
            class="MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiStepIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                class="MuiStepIcon-text"
                text-anchor="middle"
                x="12"
                y="16"
              >
                3
              </text>
            </svg>
          </span>
          <span
            class="MuiStepLabel-labelContainer"
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
